### PR TITLE
[DOC] Clarify Wasm, VMContext, Env. Add Rust Transaction example

### PR DESCRIPTION
**Pull request by RNHTTR**
**Is your feature request related to a problem? Please describe.**
It's difficult to understand the relationship between wasm, VMContext, and env, and it's really hard to find how to send NEAR from user to user. 

**Describe the solution you'd like**
A specific page for Wasm, VMContext, and env and a code example for using `Promise` to transfer funds between two users. I could use help describing wasm and vmcontext.

### CODE REVIEW

**Comment by RNHTTR** [comment]
Also should include documentation providing context around contract initialization, re-deploying a contract and what happens to state, and how to access blockchain storage

**Comment by RNHTTR** [comment]
> Also should include documentation providing context around contract initialization, re-deploying a contract and what happens to state, and how to access blockchain storage

Also some additional commentary on the idea that ["All data stored on the NEAR blockchain is done in key / value pairs."](https://docs.near.org/docs/concepts/data-storage#overview). It'd be useful to describe in `docs/concepts/data-storage` how to initialize a `Vector`, `UnorderedMap`, etc., and what it means to pass `prefix: S` / `IntoStorageKey`

File: docs/develop/contracts/rust/wasm-vmcontext-and-env.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+TODO
+
+## VMContext
+TODO
+
+## env
+[`near-sdk::env`](https://docs.rs/near-sdk/latest/near_sdk/env/index.html) makes blockchain specific methods available to the smart contract. These methods include (but aren't limited to):
+* [`account_balance`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.account_balance.html)
+* [`current_account_id`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.current_account_id.html)
+* [`is_valid_account_id`](https://docs.rs/near-sdk/latest/near_sdk/env/fn.is_valid_account_id.html)
```
Comment (austinabell) [inline docs/develop/contracts/rust/wasm-vmcontext-and-env.md:20]: why do we want to advertise this? If anything we should note that AccountId's are being validated through its type (as of v4) and not require using this API which is very unergonomic and should only be used in edge cases

File: docs/develop/contracts/rust/near-sdk-rs.md [hunk 2]
```diff
@@ -138,17 +138,22 @@ impl Default for StatusMessage {
...
+    Promise,
+}
 
 #[payable]
-pub fn my_method(&mut self) {
-...
+pub fn send_funds(&mut self) {
+    let receiver_id: AccountId = "receiver.testnet".to_string();
+    let amount: u128 = 1_000_000_000_000_000_000_000_000;
+    Promise::new(receiver_id).transfer(amount);
```
Comment (austinabell) [inline docs/develop/contracts/rust/near-sdk-rs.md:153]: I'm confused, this doesn't make sure that the necessary funds were sent to the contract, isn't this a bad pattern to use as an example?

**Comment by austinabell** [comment]
Thanks for the contribution @RNHTTR, a lot of the rust documentation is being updated/added here https://www.near-sdk.io/ which includes roughly the changes you're proposing here. https://github.com/near/sdk-docs is the repo, for reference.

**Comment by RNHTTR** [comment]
> Thanks for the contribution @RNHTTR, a lot of the rust documentation is being updated/added here https://www.near-sdk.io/ which includes roughly the changes you're proposing here. https://github.com/near/sdk-docs is the repo, for reference.

Aha! That looks great, and like you said, covers a lot of what I'm trying to capture. Ultimately, as I learn about NEAR & the rust sdk, I wanted to capture some of the things that I found challenging/confusing and be proactive in updating the documentation accordingly.

Not sure this PR is the best forum to continue discussion, but I've found the following difficult to follow:

- What is `env`, what are its properties/methods, and why are these significant?
  - Where does `env` even come from? Can these be manipulated from the front-end when the contract is called?
  - Who exactly is `env::current_account_id`, `env::predecessor_account_id`, `env::signer_account_id`, and how are these set?
- How can a contract let user A send NEAR to user B? This is what I was trying to capture with the `send_funds` method, but it looks what I was trying to convey was nicely captured [in the docs you shared](https://www.near-sdk.io/promises/token-tx).
- Understanding the on-chain storage components of a contract. I thought that when a contract was deployed, each time the contract was called, the `new` method would be called, and the properties in `new` would be set each time the contract was called.
  - On a similar note, seeing code like `LookupMap::new(b"my_property".to_vec())` was really confusing to me. It'd be nice if there was some additional clarification on the key / value paradigm for on-chain storage, outlined in the [overview here](https://docs.near.org/docs/concepts/data-storage#overview). Now I understand that `"my_property".to_vec()` sets the key for the storage object on-chain. At least I think I understand it :)

Let me know if this makes sense or not. Thanks!

**Comment by austinabell** [comment]
> Not sure this PR is the best forum to continue discussion, but I've found the following difficult to follow:
To

> 
> * What is `env`, what are its properties/methods, and why are these significant?
>   
>   * Where does `env` even come from? Can these be manipulated from the front-end when the contract is called?
>   * Who exactly is `env::current_account_id`, `env::predecessor_account_id`, `env::signer_account_id`, and how are these set?
> * How can a contract let user A send NEAR to user B? This is what I was trying to capture with the `send_funds` method, but it looks what I was trying to convey was nicely captured [in the docs you shared](https://www.near-sdk.io/promises/token-tx).
> * Understanding the on-chain storage components of a contract. I thought that when a contract was deployed, each time the contract was called, the `new` method would be called, and the properties in `new` would be set each time the contract was called.
>   
>   * On a similar note, seeing code like `LookupMap::new(b"my_property".to_vec())` was really confusing to me. It'd be nice if there was some additional clarification on the key / value paradigm for on-chain storage, outlined in the [overview here](https://docs.near.org/docs/concepts/data-storage#overview). Now I understand that `"my_property".to_vec()` sets the key for the storage object on-chain. At least I think I understand it :)
> 
> Let me know if this makes sense or not. Thanks!

**Comment by RNHTTR** [comment]
I think it's okay to close. In my opinion, having some of this info up front would make the learning curve a little gentler. Just food for thought :)

Thanks for the explanations!

**Pull request closed without merging**